dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-zeppelin/src/test/java/org/apache/dolphinscheduler/plugin/datasource/zeppelin/ZeppelinDataSourceProcessorTest.java

```diff
@@ -0,0 +1,106 @@
+                () -> zeppelinDataSourceProcessor.checkDatasourceParam(zeppelinDataSourceParamDTO));
+        zeppelinDataSourceParamDTO.setRestEndpoint("http://dolphinscheduler.com:8080");
+        Assertions.assertThrows(IllegalArgumentException.class,
+                () -> zeppelinDataSourceProcessor.checkDatasourceParam(zeppelinDataSourceParamDTO));
+        zeppelinDataSourceParamDTO.setUserName("root");
+        Assertions
+                .assertDoesNotThrow(() -> zeppelinDataSourceProcessor.checkDatasourceParam(zeppelinDataSourceParamDTO));
+    }
+
+    @Test
+    void testGetDatasourceUniqueId() {
+        ZeppelinConnectionParam zeppelinConnectionParam = new ZeppelinConnectionParam();
+        zeppelinConnectionParam.setRestEndpoint("https://dolphinscheduler.com:8080");
+        zeppelinConnectionParam.setUsername("root");
+        zeppelinConnectionParam.setPassword("123456");
+        Assertions.assertEquals("zeppelin@https://dolphinscheduler.com:8080@root@123456",
+                zeppelinDataSourceProcessor.getDatasourceUniqueId(zeppelinConnectionParam, DbType.ZEPPELIN));
+
+    }
+
+    @Test
+    void testCreateDatasourceParamDTO() {
+        ZeppelinDataSourceParamDTO zeppelinDataSourceParamDTO =
+                (ZeppelinDataSourceParamDTO) zeppelinDataSourceProcessor.createDatasourceParamDTO(connectJson);
+        Assertions.assertEquals("lucky", zeppelinDataSourceParamDTO.getUserName());
+        Assertions.assertEquals("123456", zeppelinDataSourceParamDTO.getPassword());
+        Assertions.assertEquals("https://dolphinscheduler.com:8080", zeppelinDataSourceParamDTO.getRestEndpoint());
+    }
+
+    @Test
+    void testCreateConnectionParams() {
+        ZeppelinDataSourceParamDTO zeppelinDataSourceParamDTO =
+                (ZeppelinDataSourceParamDTO) zeppelinDataSourceProcessor.createDatasourceParamDTO(connectJson);
+        ZeppelinConnectionParam zeppelinConnectionParam =
+                zeppelinDataSourceProcessor.createConnectionParams(zeppelinDataSourceParamDTO);
+        Assertions.assertEquals("lucky", zeppelinConnectionParam.getUsername());
+        Assertions.assertEquals("123456", zeppelinConnectionParam.getPassword());
+        Assertions.assertEquals("https://dolphinscheduler.com:8080", zeppelinConnectionParam.getRestEndpoint());
+    }
+
+    @Test
+    void testTestConnection() throws Exception {
+        ZeppelinDataSourceParamDTO zeppelinDataSourceParamDTO =
+                (ZeppelinDataSourceParamDTO) zeppelinDataSourceProcessor.createDatasourceParamDTO(connectJson);
+        ZeppelinConnectionParam zeppelinConnectionParam =
+                zeppelinDataSourceProcessor.createConnectionParams(zeppelinDataSourceParamDTO);
+
+        Mockito.doNothing().when(zeppelinClient).login(zeppelinConnectionParam.getUsername(),
+                zeppelinConnectionParam.getPassword());
+        boolean isConnected = zeppelinDataSourceProcessor.testConnection(zeppelinConnectionParam);
```

Comment:
The reason why this case fails is that you explicitly use the `new` keyword to instantiate the `ZeppelinClient` in `ZeppelinDataSourceProcessor`. When you do the stubbing here, the mocked `zeppelinClient` object is not injected into the `zeppelinDataSourceProcessor` thus the stubbing takes no effect. The best practice is to decouple the process of instantiation of `ZeppelinClient` from `ZeppelinDataSourceProcessor`.
![image](https://github.com/apache/dolphinscheduler/assets/34905992/c503c352-1343-4825-9694-5033d9402f6b)

Reply:
Thank you very much for your help